rl4co/envs/routing/pdp/env.py

```diff
@@ -209,3 +210,344 @@ def check_solution_validity(self, td, actions):
+    def _step(cls, td: TensorDict) -> TensorDict:
+        # get state information from td
+        action = td["action"]
+        selected = action[:, 0].view(-1, 1)
+        first = action[:, 1].view(-1, 1)
+        second = action[:, 2].view(-1, 1)
+        solution = td["rec_current"]
+        solution_best = td["rec_best"]
+        locs = td["locs"]
+        cost_bsf = td["cost_bsf"]
+        action_record = td["action_record"]
+        bs, gs = solution.size()
+
+        # perform ruin and repair
+        next_rec = cls._insert_operator(solution, selected + 1, first, second)
+        new_obj = cls.get_costs(locs, next_rec)
+
+        # compute reward and update best-so-far solutions
+        now_bsf = torch.where(new_obj < cost_bsf, new_obj, cost_bsf)
+        reward = cost_bsf - now_bsf
+        index = reward > 0.0
+        solution_best[index] = next_rec[index].clone()
+
+        # reset visited_time
+        visited_time = td["visited_time"] * 0
+        pre = torch.zeros((bs), device=visited_time.device).long()
+        arange = torch.arange(bs)
+        for i in range(gs):
+            current_nodes = next_rec[arange, pre]
+            visited_time[arange, current_nodes] = i + 1
+            pre = next_rec[arange, pre]
+        visited_time = visited_time.long()
+
+        # update action record
+        action_record[:, :-1] = action_record[:, 1:]
+        action_record[:, -1] *= 0
+        action_record[torch.arange(bs), -1, action[:, 0]] = 1
+
+        # Update step
+        td.update(
+            {
+                "cost_current": new_obj,
+                "cost_bsf": now_bsf,
+                "rec_current": next_rec,
+                "rec_best": solution_best,
+                "visited_time": visited_time,
+                "action_record": action_record,
+                "i": td["i"] + 1,
+                "reward": reward,
+            }
+        )
+
+        return td
+
+    def _reset(self, td: Optional[TensorDict] = None, batch_size=None) -> TensorDict:
+        device = td.device
+
+        locs = torch.cat((td["depot"][:, None, :], td["locs"]), -2)
+        current_rec = self.generator._get_initial_solutions(locs).to(device)
+
+        obj = self.get_costs(locs, current_rec)
+
+        # get index according to the solutions in the linked list data structure
+        bs = batch_size[0]
+        seq_length = self.generator.num_loc + 1
+        visited_time = torch.zeros((bs, seq_length), device=device)
+        pre = torch.zeros((bs), device=device).long()
+        arange = torch.arange(bs)
+        for i in range(seq_length):
+            current_nodes = current_rec[arange, pre]
+            visited_time[arange, current_nodes] = i + 1
+            pre = current_rec[arange, pre]
+        visited_time = visited_time.long()
+
+        # get action record and step i
+        i = torch.zeros((*batch_size, 1), dtype=torch.int64, device=device)
+        action_record = torch.zeros((bs, seq_length // 2, seq_length // 2))
```

Comment:
[Minor] Maybe move to device for safe init.

Reply:
Okay, changed accordingly!